### DISCUSSION TITLE
image bright/dark part fx (for exposure bracket clipping detection)

### DISCUSSION THREAD

**Discussion by gitoss** [ImageMagick/ImageMagick#3592] [Help] [answered]
I'm using 5x aperture bracketing with either my dslr or my mobile phone camera - just to be on the safe side if the automatic exposure misses or the dynamic range is exceeded.

I've written a batch script to automatically detect images in the bracket that are clipped, and delete them if another image in the bracket isn't clipped. I'm using fx:minima and fx:maxima for this.  However, very often _all_ images of the bracket are clipped - for example if the sun and/or a a very deep shadow is in the frame. In this case, fx:minima is aways 0, and fx:maxima is always 1 in all brackets. 

To get around this problem, I'd like to **calculate what percentage/area of the image is 0 or 1** (sometimes dcraw seems to miss reconstucting the raw image, so I have to use **<0.1 and >0.9** respectively). _**Question:** What's the fx for this with ImageMagick?_

Rationale: If the very bright or very dark part is nearly identical in multiple bracket images, it's very likely that it is severely clipped and hdr-merging wouldn't help.

**Comment by snibgo** [comment, marked as answer]
What version of IM, on what platform? I will assume IM v7 on Windows BAT.

Clipping at either end can occur in any of the RGB channels. If we have W\*H pixels, there are W\*H\*3 opportunities for clipping at each end.

Put this in a BAT script named clippedArea.bat:
```
rem %1 input image

magick ^
  %1 ^
  -channel RGB -separate +channel +append +repage ^
  ( +clone ^
    -threshold 99%% ^
    -format "HI=%%[fx:mean*w*h]\n" ^
    +write info: ^
    +delete ^
  ) ^
  ( +clone ^
    -threshold 1%% ^
    -format "LO=%%[fx:(1-mean)*w*h]\n" ^
    +write info: ^
    +delete ^
  ) ^
  NULL:
```
For example:
```
f:\web\im>clippedArea.bat toes.png

HI=23
LO=1
```
With these limits of 99% at the top and 1% at the bottom, 23 pixel values (out of W\*H\*3) have clipped at the top, and 1 has clipped at the bottom.

In the command, we separate and append the three channels, giving us a grayscale image that is 3 times the size of the input.

From the "-format" options you can remove "\*w\*h" to get a proportion of the pixel values that have clipped. Or multiply by 100 to get a percentage.

**Reply by gitoss** [reply]
That's it, I didn't know the "threshold" option yet - Thanks!

**Comment by Jmuccigr** [comment]
You **should** have clipping in some of your bracketed photos. The idea of the process is to get the darkest portions light enough in the brightest photo and the lightest portions dark enough in the darkest photo. Then your software will choose the appropriate portions from the set of photos when it does the merge.

**Reply by gitoss** [reply]
Thanks, I have a general idea what bracketing does :-)

**Reply by Jmuccigr** [reply]
Great, so then you should know that clipping is not only expected, but desired in the extreme images in the series.

**Reply by gitoss** [reply]
Of course if your dslr sensor is 13 years old like mine, in some cases clipping isn't desired but a healthy distance to the edges of the histogram - because of sensor noise in the shadows, and too fragile gradient and color bleed (Adobe engine) when recovering highlights from the raw file.

**Comment by snibgo** [comment]
I agree with @Jmuccigr. When I take a series of bracketed photos, I expect at the one with greatest exposure will have badly clipped highlights, and that exposure will show the best detail in the shadows. The HDR (High Dynamic Range) processing should use that shadow detail, and ignore clipped pixels.

Similarly, the lowest-exposure photo will have clipped (or "crushed") shadows, but best detail in the highlights.

EDIT: Of course, if one of the bracketed series has no clipping, then the scene was within the dynamic range of the camera. But we often have specular highlights that clip, and a test such as the above can be used to test whether a _significant proportion_ of the pixels have clipped.

**Reply by gitoss** [reply]
The above detection is a) to detect if one bracket is nearly inside the dynamic range of the camera (i.e. I can dump the other brackets), or to b) detect if extreme brackets are unneccessary.

For safety to guard against wrong auto exposure, I'm mostly taking 5x brackets of everything so I need to automatically filter the brackets that I can dump right away, without any need to import them into Lightroom and manual selection.

For example, if the sun is in the frame, or a very dark shadow (like an open doorway to a pitch black room) these spots will never be inside the dynamic range. So I need to detect the best "necessary" bracket for example in a sequence -4ev, -2ev, 0ev, +2ev, +4ev.

Very often, the 4ev brackets won't have any additional usable information that's not already inside the 2ev brackets - just the histogram was shifted, and there's just more "space" to the left or right. Plus in the +4v bracket, there's added danger of motion blur or (handheld) camera shake because of the longer exposure time (even if it might have a little less digital noise in the shadows), and the -4ev bracket will have more digital noise (because everything is shifted to the left).

Last not least, even if there is just a tiny little bit of more information in the +-4ev brackets, in my experience it ususally doesn't show in a 5x hdr fusion vs 3x - but requires more storage for the raw files, and more processing time. If the +-4 ev brackets have blur/shake or noise, the 5x brackets might even turn out worse.

**Reply by Jmuccigr** [reply]
I'm not so sure you can easily decide that highly clipped images don't contain useful information. My experience with heavy shadowed images is that it takes a lot of bright-end clipping to get an image that shows detail in the shadows.

Shake is a different issue, of course.

**Reply by gitoss** [reply]
That's correct, and I'm acutally only deleting the very severe cases (nearly white/black picture) right away.

When in doubt, I'm keyword/label-marking the images with exiftool, so I can easily filter and recognize (by colour label) in Lightroom which images are clipped or mostly inside the dynmic range, and then sort through them manually and hdr-merge when necessary.

I'd like to have a more automated approach, but as you've written it seems to depend on the content rather than just on the histogram. But at least a first-pass purge and then marking the brackets is a relief.
